s2s/time.py

```diff
@@ -170,6 +173,48 @@ def map_years(
 
         return index
 
+    def map_to_data(self, input_data: Union[pd.Series, pd.DataFrame, xr.Dataset, xr.DataArray],
+        flat: bool = False
+        ) -> pd.DataFrame:
+        """Map the calendar to input data period.
+        
+        Get datetime range from input data and generate corresponding interval index. This method
+        guarentees that the generated interval (calendar) indices would be covered by the input
+        data.
+        Args:
+            input_data: Input data for datetime mapping. Its index must be pandas.DatetimeIndex.
+            flat: Same as the argument in ``map_years``.
+        Returns:
+            Pandas DataFrame filled with Intervals of the calendar's frequency.
+            (also see ``map_year`` and ``map_years``)
+        """
+        # check the datetime order of input data
+        first_timestamp = input_data.index.min()
+        last_timestamp = input_data.index.max()
+
+        map_last_year = last_timestamp.year
+        # ensure that the input data could always cover the advent calendar
+        map_first_year = first_timestamp.year + 1
+        # check if the last date(time) is covered by the advent calendar
+        anchor_date_with_year = pd.Timestamp(year=map_last_year, month=self.month, day=self.day)
+        if anchor_date_with_year > last_timestamp:
+            map_last_year = map_last_year - 1
+
+        if map_last_year > map_first_year:
+            intervals = self.map_years(map_first_year, map_last_year, flat)
+            # check if the input data cover all the intervals
+            if intervals.iloc[-1,-1].left < first_timestamp:
+                warnings.warn("The last few intervals are not fully covered by the input data.", UserWarning)
+        elif map_last_year == map_first_year:
+            intervals = self.map_year(map_last_year)
+            # check if the input data cover all the intervals
+            if intervals.iloc[-1].left < first_timestamp:
+                warnings.warn("The last few intervals are not fully covered by the input data.", UserWarning)
```

Comment:
This line of code is repeated and can be taken out of the if block. Also, not sure if this is still needed, since you now enforce that the input will be entirely covered

Reply:
If we take the code out of this block, then we need to make extra check for the dimension of intervals to refer to the last element (advent calendar can be 2d or flat in 1d), which is not so nice. I prefer to keep it this way for now. It is needed only temporarily. When we implement the `discard` function with `max_lag` and trim the calendar, we know exactly what the calendar looks like and can then enforce that the input will be entirely covered (currently the incomplete intervals are allowed, e.g. the `AdventCalendar` goes with `archor date = (10, 15)` and `freq='30d'`, the input data is from 2021.11.10 - 2022.12.15, 10 days frequency, then the last interval "(2020-10-20, 2020-11-19]" is slightly incomplete, but we need to warn the user).

This also leads to another issue #27 , about the order of calling these methods (e.g. do we expect our user to instantiate `AdventCalendar` first, map to data, then discard intervals based on `max_lag`?). I have discussed it with @BSchilperoort and opened an issue for this, let's bring our discussion there.

Reply:
In that case, can you write a test that triggers this warning?

Reply:
Reviewing this from my phone, so couldn't test it, but shouldn't the iloc be 0 instead of -1?

Reply:
Based on the discussion in #27 , this problem is solved then. We don't need those warnings now as we ensure that the input will be entirely covered.